File: src/main/java/com/springdebugger/classifier/RuleBasedClassifier.java
package com.springdebugger.classifier;

import com.springdebugger.engine.DiagnosisEngine;
import com.springdebugger.model.Confidence;
import com.springdebugger.model.DiagnosisCard;
import com.springdebugger.model.RawSignal;
import com.springdebugger.rule.Rule;
import com.springdebugger.rule.RuleCatalog;
import com.springdebugger.rule.SignalCriteria;

import java.util.Optional;

/**
 * Matches a RawSignal against the loaded rule catalog and produces a DiagnosisCard.
 * Rules are evaluated in catalog order; the first rule where all non-null criteria match wins.
 *
 * This is the offline rule engine: the first {@link DiagnosisEngine} the pipeline tries.
 */
public final class RuleBasedClassifier implements DiagnosisEngine {

    private final RuleCatalog catalog;

    public RuleBasedClassifier(RuleCatalog catalog) {
        this.catalog = catalog;
    }

    @Override
    public Optional<DiagnosisCard> diagnose(RawSignal signal) {
        return classify(signal);
    }

    public Optional<DiagnosisCard> classify(RawSignal signal) {
        for (Rule rule : catalog.all()) {
            // Only validated rules are active. A rule is DONE only when it has a
            // fixture that passes ClassifierFixtureTest; TODO rules are unvalidated
            // and must not fire in production.
            if (!"DONE".equals(rule.getStatus())) {
                continue;
            }
            if (rule.getPhases() != null && !rule.getPhases().isEmpty()
                    && !rule.getPhases().contains(signal.getPhase())) {
                continue;
            }
            if (matches(rule.getSignals(), signal)) {
                String diagnosis = fillTemplate(rule.getDiagnosis(), signal, rule);
                String fix = fillTemplate(rule.getFix(), signal, rule);
                return Optional.of(new DiagnosisCard(
                        rule.getId(),
                        signal.getPhase(),
                        diagnosis,
                        fix,
                        rule.getConfidence(),
                        signal.getRawExcerpt()));
            }
        }
        return Optional.empty();
    }

    private boolean matches(SignalCriteria criteria, RawSignal signal) {
        if (criteria == null) return false;

        if (criteria.getCausedByClass() != null) {
            if (signal.getDeepestCausedByClass() == null) return false;
            if (!containsIgnoreCase(signal.getDeepestCausedByClass(), criteria.getCausedByClass())) return false;
        }

        if (criteria.getCausedByMessage() != null) {
            if (signal.getDeepestCausedByMessage() == null) return false;
            if (!containsIgnoreCase(signal.getDeepestCausedByMessage(), criteria.getCausedByMessage())) return false;
        }

        if (criteria.getMessageContains() != null) {
            boolean inLines = signal.anyLineContains(criteria.getMessageContains());
            boolean inExcerpt = containsIgnoreCase(signal.getRawExcerpt(), criteria.getMessageContains());
            if (!inLines && !inExcerpt) return false;
        }

        if (criteria.getBannerDescriptionContains() != null) {
            if (signal.getBannerDescription() == null) return false;
            if (!containsIgnoreCase(signal.getBannerDescription(), criteria.getBannerDescriptionContains())) return false;
        }

        if (criteria.getBannerActionContains() != null) {
            if (signal.getBannerAction() == null) return false;
            if (!containsIgnoreCase(signal.getBannerAction(), criteria.getBannerActionContains())) return false;
        }

        if (criteria.getHttpStatus() > 0) {
            if (signal.getHttpStatus() != criteria.getHttpStatus()) return false;
        }

        if (criteria.getBuildLineContains() != null) {
            boolean inLines = signal.anyLineContains(criteria.getBuildLineContains());
            boolean inExcerpt = containsIgnoreCase(signal.getRawExcerpt(), criteria.getBuildLineContains());
            if (!inLines && !inExcerpt) return false;
        }

        return true;
    }

    private static final java.util.regex.Pattern PORT =
            java.util.regex.Pattern.compile("[Pp]ort[\\s:]+(\\d{2,5})");
    private static final java.util.regex.Pattern PLACEHOLDER =
            java.util.regex.Pattern.compile("Could not resolve placeholder ['\"]?([\\w.\\-]+)");

    private String fillTemplate(String template, RawSignal signal, Rule rule) {
        if (template == null) return "";
        String result = template;
        if (result.contains("{{beanType}}")) {
            result = result.replace("{{beanType}}", signal.getDeepestCausedByMessage() != null
                    ? extractTypeName(signal.getDeepestCausedByMessage()) : "the required type");
        }
        if (result.contains("{{beanName}}")) {
            result = result.replace("{{beanName}}", signal.getFailingBeanName() != null
                    ? signal.getFailingBeanName() : "the failing bean");
        }
        if (result.contains("{{port}}")) {
            result = result.replace("{{port}}", firstMatch(PORT, signal, "the configured port"));
        }
        if (result.contains("{{property}}")) {
            result = result.replace("{{property}}", firstMatch(PLACEHOLDER, signal, "the property"));
        }
        result = result.replace("{{ruleId}}", rule.getId());
        return result;
    }

    /** First capture group of the pattern across the deepest message, banner, then excerpt. */
    private String firstMatch(java.util.regex.Pattern pattern, RawSignal signal, String fallback) {
        for (String text : new String[]{signal.getDeepestCausedByMessage(),
                signal.getBannerDescription(), signal.getRawExcerpt()}) {
            if (text == null) continue;
            java.util.regex.Matcher m = pattern.matcher(text);
            if (m.find()) return m.group(1);
        }
        return fallback;
    }

    private static final java.util.regex.Pattern QUOTED_TYPE =
            java.util.regex.Pattern.compile("'([\\w.$]+)'");
    private static final java.util.regex.Pattern OF_TYPE =
            java.util.regex.Pattern.compile("of type \\[?([\\w.$]+)");

    /**
     * Extracts the bean type name from an exception message, e.g. from
     * "No qualifying bean of type 'com.example.FgClassifier' available" returns "FgClassifier".
     * Prefers a quoted fully-qualified name, then an "of type X" form.
     */
    private String extractTypeName(String message) {
        java.util.regex.Matcher q = QUOTED_TYPE.matcher(message);
        while (q.find()) {
            String candidate = q.group(1);
            if (candidate.contains(".") || (!candidate.isEmpty() && Character.isUpperCase(candidate.charAt(0)))) {
                return simpleName(candidate);
            }
        }
        java.util.regex.Matcher o = OF_TYPE.matcher(message);
        if (o.find()) return simpleName(o.group(1));
        return "the required type";
    }

    private String simpleName(String fqn) {
        int dot = fqn.lastIndexOf('.');
        return dot >= 0 ? fqn.substring(dot + 1) : fqn;
    }

    private boolean containsIgnoreCase(String text, String substring) {
        return text.toLowerCase().contains(substring.toLowerCase());
    }
}

File: vscode-extension/src/engine/rule-based-classifier.ts
// Port of com.springdebugger.classifier.RuleBasedClassifier. Rules are evaluated in catalog order;
// the first DONE rule whose phase matches and whose every present criterion matches wins. Template
// placeholders are filled from the signal exactly as in Java.
import { DiagnosisCard, RawSignal } from './models';
import { RuleCatalog } from './rule-catalog';
import { Rule, SignalCriteria } from './rule';

const PORT = /[Pp]ort[\s:]+(\d{2,5})/;
const PLACEHOLDER = /Could not resolve placeholder ['"]?([\w.\-]+)/;
const OF_TYPE = /of type \[?([\w.$]+)/;

export class RuleBasedClassifier {
  constructor(private readonly catalog: RuleCatalog) {}

  classify(signal: RawSignal): DiagnosisCard | null {
    for (const rule of this.catalog.all()) {
      // Only validated (DONE) rules are active; TODO rules must not fire in production.
      if (rule.status !== 'DONE') continue;
      if (
        rule.phases !== null &&
        rule.phases.length > 0 &&
        !rule.phases.includes(signal.phase)
      ) {
        continue;
      }
      if (this.matches(rule.signals, signal)) {
        return {
          ruleId: rule.id,
          phase: signal.phase,
          diagnosisSentence: this.fillTemplate(rule.diagnosis, signal, rule),
          fixSentence: this.fillTemplate(rule.fix, signal, rule),
          confidence: rule.confidence,
          excerpt: signal.rawExcerpt,
        };
      }
    }
    return null;
  }

  private matches(criteria: SignalCriteria | null, signal: RawSignal): boolean {
    if (criteria === null) return false;
    // A rule with no active signal must never match (it would match everything).
    if (!hasAnySignal(criteria)) return false;

    if (criteria.causedByClass != null) {
      if (signal.deepestCausedByClass === null) return false;
      if (!containsIgnoreCase(signal.deepestCausedByClass, criteria.causedByClass)) return false;
    }

    if (criteria.causedByMessage != null) {
      if (signal.deepestCausedByMessage === null) return false;
      if (!containsIgnoreCase(signal.deepestCausedByMessage, criteria.causedByMessage)) return false;
    }

    if (criteria.messageContains != null) {
      const inLines = signal.anyLineContains(criteria.messageContains);
      const inExcerpt = containsIgnoreCase(signal.rawExcerpt, criteria.messageContains);
      if (!inLines && !inExcerpt) return false;
    }

    if (criteria.messageContainsAny != null && criteria.messageContainsAny.length > 0) {
      let anyMatch = false;
      for (const alt of criteria.messageContainsAny) {
        if (signal.anyLineContains(alt) || containsIgnoreCase(signal.rawExcerpt, alt)) {
          anyMatch = true;
          break;
        }
      }
      if (!anyMatch) return false;
    }

    if (criteria.bannerDescriptionContains != null) {
      if (signal.bannerDescription === null) return false;
      if (!containsIgnoreCase(signal.bannerDescription, criteria.bannerDescriptionContains)) return false;
    }

    if (criteria.bannerActionContains != null) {
      if (signal.bannerAction === null) return false;
      if (!containsIgnoreCase(signal.bannerAction, criteria.bannerActionContains)) return false;
    }

    if ((criteria.httpStatus ?? 0) > 0) {
      if (signal.httpStatus !== criteria.httpStatus) return false;
    }

    if (criteria.buildLineContains != null) {
      const inLines = signal.anyLineContains(criteria.buildLineContains);
      const inExcerpt = containsIgnoreCase(signal.rawExcerpt, criteria.buildLineContains);
      if (!inLines && !inExcerpt) return false;
    }

    return true;
  }

  private fillTemplate(template: string | null, signal: RawSignal, rule: Rule): string {
    if (template === null) return '';
    let result = template;
    if (result.includes('{{beanType}}')) {
      const source =
        signal.deepestCausedByMessage !== null
          ? signal.deepestCausedByMessage
          : signal.bannerDescription;
      result = result.split('{{beanType}}').join(source !== null ? extractTypeName(source) : 'the required type');
    }
    if (result.includes('{{beanName}}')) {
      result = result
        .split('{{beanName}}')
        .join(signal.failingBeanName !== null ? signal.failingBeanName : 'the failing bean');
    }
    if (result.includes('{{port}}')) {
      result = result.split('{{port}}').join(firstMatch(PORT, signal, 'the configured port'));
    }
    if (result.includes('{{property}}')) {
      result = result.split('{{property}}').join(firstMatch(PLACEHOLDER, signal, 'the property'));
    }
    result = result.split('{{ruleId}}').join(rule.id);
    return result;
  }
}

/** First capture group of the pattern across the deepest message, banner, then excerpt. */
function firstMatch(pattern: RegExp, signal: RawSignal, fallback: string): string {
  for (const text of [signal.deepestCausedByMessage, signal.bannerDescription, signal.rawExcerpt]) {
    if (text === null) continue;
    const m = pattern.exec(text);
    if (m) return m[1];
  }
  return fallback;
}

/**
 * Extracts the bean type name from an exception message. Prefers a quoted fully-qualified name,
 * then an "of type X" form. Mirrors RuleBasedClassifier.extractTypeName.
 */
function extractTypeName(message: string): string {
  const quoted = /'([\w.$]+)'/g;
  let q: RegExpExecArray | null;
  while ((q = quoted.exec(message)) !== null) {
    const candidate = q[1];
    if (candidate.includes('.') || (candidate.length > 0 && isUpperCase(candidate.charAt(0)))) {
      return simpleName(candidate);
    }
  }
  const o = OF_TYPE.exec(message);
  if (o) return simpleName(o[1]);
  return 'the required type';
}

function simpleName(fqn: string): string {
  const dot = fqn.lastIndexOf('.');
  return dot >= 0 ? fqn.substring(dot + 1) : fqn;
}

/** Mirrors Character.isUpperCase for the cased letters that appear in class names. */
function isUpperCase(ch: string): boolean {
  return ch !== ch.toLowerCase() && ch === ch.toUpperCase();
}

function hasAnySignal(c: SignalCriteria): boolean {
  return (
    c.causedByClass != null ||
    c.causedByMessage != null ||
    c.messageContains != null ||
    (c.messageContainsAny != null && c.messageContainsAny.length > 0) ||
    c.bannerDescriptionContains != null ||
    c.bannerActionContains != null ||
    c.buildLineContains != null ||
    c.exceptionClass != null ||
    (c.httpStatus ?? 0) > 0
  );
}

function containsIgnoreCase(text: string, substring: string): boolean {
  return text.toLowerCase().includes(substring.toLowerCase());
}
